Added Styling & Rendering sections.

arthurevans opened the pull request:
Fixes #711 and #710.

Staged at https://styling-updates-dot-polymer-lit-html.appspot.com

arthurevans commented:
Sorry for the very late arrival. Something to read on the plane.

Tomorrow: Tooling!

justinfagnani commented on docs/_guide/03-styling-templates.md:
```diff
@@ -0,0 +1,94 @@
+slug: styling-templates
+---
+
+lit-html just renders HTML; you can style HTML created with lit-html like any other HTML, using CSS. How you apply styles to lit-html templates depends on how you're using it—for example, if you're using lit-html inside a component system like LitElement, you can follow the patterns used by that component system.
```
```suggestion
lit-html focuses on one thing: rendering HTML. How you apply styles to the HTML lit-html creates depends on how you're using it—for example, if you're using lit-html inside a component system like LitElement, you can follow the patterns used by that component system.
```

justinfagnani commented on docs/_guide/03-styling-templates.md:
```diff
@@ -0,0 +1,94 @@
+
+lit-html just renders HTML; you can style HTML created with lit-html like any other HTML, using CSS. How you apply styles to lit-html templates depends on how you're using it—for example, if you're using lit-html inside a component system like LitElement, you can follow the patterns used by that component system.
+
+In most cases you should **not** use lit-html bindings inside a style sheet. It's more efficient to use lit-html bindings to manipulate the `class` and `style` attributes.  
```
Seems like a little jump from the intro para to here, and there's a "Don't use bindings in stylesheets" section below. Does this belong there?

justinfagnani commented on docs/_guide/03-styling-templates.md:
```diff
@@ -0,0 +1,94 @@
+
+See [classMap](template-reference#classmap) and [styleMap](template-reference#stylemap) in the Template syntax reference.
+
+There are also some special considerations when using lit-html to render into a shadow root.
```
I think rather than calling Shadow DOM as special, maybe divide the world into two: you're either going to be using Shadow DOM or not.

If not, you're rendering to the main doc and you should use whatever system you already do. A simple stylesheet or various CSS module / CSS-in-JS systems. If you use those you might need to bind class names (it's common for a CSS module system to have you import class names to put on elements), and that's easy in lit-html. For more information see your CSS libraries docs, etc.

If you are using Shadow DOM (recommended), then... (what you have below)

justinfagnani commented on docs/_guide/03-styling-templates.md:
````diff
@@ -0,0 +1,94 @@
+```
+
+
+### Don't use bindings in stylesheets 
````
I would consider renaming this "Binding in Stylesheets"

Then recommending that in most cases you don't. In native Shadow DOM it's not that bad if the stylesheet is very small, for instance:

```js
html`
<style>
:host {
  --my-special-color: ${specialColor}
}
</style>
`
```

Although, if you're templated to do this, `styleMap` is probably better, so you could show that there.

justinfagnani commented on docs/_guide/03-styling-templates.md:
```diff
@@ -0,0 +1,94 @@
+
+In most cases you should **not** use lit-html bindings inside a style sheet. It's more efficient to use lit-html bindings to manipulate the `class` and `style` attributes.  
+
+lit-html provides two directives for manipulating an element's `class` and `style` attributes:
```
I think styleMap and classMap deserve a heading here. Maybe: "Styling individual Elements with classMap and styleMap". An example could be useful too.

arthurevans replied:
Yeah, I hated to add a section here because it felt really repetitive with the section in the reference. 

That example is already incredibly stupid, and I don't really want to repeat it ... Would love suggestions for better examples here.

katejeffreys replied:
https://github.com/katejeffreys/classmaps/blob/master/index.js has a couple i was playing with

justinfagnani commented on docs/_guide/03-styling-templates.md:
```diff
@@ -0,0 +1,94 @@
+
+## Rendering in shadow DOM
+
+When rendering into a shadow root, you usually want to add a stylesheet inside the shadow root to the template, to you can style the contents of the shadow root. 
```
I'd call out that inline styles are perfectly OK in ShadowRoots because browsers optimize this case, then maybe mention that using `adoptedStylesheets` can be even better, and link to LitElement docs (do we have docs on that yet?)

arthurevans replied:
OK, here you're using "inline style" to refer to a style sheet in a shadow root, right?

arthurevans replied:
OK, fixed.

justinfagnani commented on docs/_guide/03-styling-templates.md:
```diff
@@ -0,0 +1,94 @@
+
+If you're using shadow DOM, you'll probably need to use polyfills to support older browsers that don't implement shadow DOM natively. [ShadyDOM](https://github.com/webcomponents/shadydom) and [ShadyCSS](https://github.com/webcomponents/shadycss) are polyfills, or shims, that emulate shadow DOM isolation and style scoping. 
+
+The lit-html `shady-render` module provides necessary integration with the shady CSS shim. If you're writing your own custom element base class that uses lit-html and shadow DOM, you'll need to use `shady-render` and also take some steps on your own. 
```
```suggestion
The lit-html `shady-render` module provides necessary integration with the ShadyCSS shim. If you're writing your own Custom Element base class that uses lit-html and Shadow DOM, you'll need to use `shady-render` and also take some steps on your own to enable it.
```

arthurevans replied:
We don't capitalize shadow DOM or custom elements in docs.

justinfagnani replied:
We do here: https://www.html5rocks.com/en/tutorials/webcomponents/shadowdom-201/
And MDN does in their text (but not in the title?): https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM

justinfagnani commented on docs/_guide/03-styling-templates.md:
````diff
@@ -0,0 +1,94 @@
+    ```js
+    import { render, TemplateResult } from 'lit-html/lib/shady-render';
+
+    render(this.myTemplate(), this.shadowRoot, { scopeName: this.tagName.toLowerCase() });
````
The `this` is kind of floating here, maybe put a class around the example and put the call in a dummy example method?

arthurevans replied:
I was trying to avoid writing an entire custom element here, because I know there are other wrinkles we're missing... I dunno... Maybe if we make it clear there are lots of things missing from the sample?

arthurevans replied:
OK, added just the skeleton of a class here.

justinfagnani commented on docs/_guide/04-rendering-templates.md:
```diff
@@ -0,0 +1,30 @@
+
+The `render` method also takes an `options` argument that allows you to specify:
+
+*   `eventContext`: The `this` value to use when invoking event listeners registered with the `@eventName` syntax.
```
eventContext only effects listener functions. Listener objects with a `handleEvent` method are always called with the listener object as the `this` value.

arthurevans replied:
Fixed, thanks.

justinfagnani commented on docs/_guide/04-rendering-templates.md:
```diff
@@ -0,0 +1,30 @@
+
+*   `eventContext`: The `this` value to use when invoking event listeners registered with the `@eventName` syntax.
+
+*   `templateFactory`: The `TemplateFactory` to use. A `TemplateFactory` creates a template element from a `TemplateResult`, typically caching templates based on their static content. Users won't usually supply their own `TemplateFactory`, but libraries that use lit-html may implement custom template factories to customize template handling.
```
maybe add "this is an advanced option"

arthurevans commented:
Addressed feedback and updated staged version at: https://styling-updates-dot-polymer-lit-html.appspot.com

Think I'm ready to merge this, but going to leave it open for another hour or two in case anyone wants to review staged version before I merge.
